tests/workflow_test.go

```diff
@@ -127,6 +129,288 @@ func (s *WorkflowTestSuite) TestStartWorkflowExecution() {
+					`,
+					historyEvents,
+				)
+
+				var wfOptionsUpdated *historypb.HistoryEvent
+				for _, e := range historyEvents {
+					if e.GetEventType() == enumspb.EVENT_TYPE_WORKFLOW_EXECUTION_OPTIONS_UPDATED {
+						wfOptionsUpdated = e
+						break
+					}
+				}
+				s.NotNil(wfOptionsUpdated)
+
+				attributes := wfOptionsUpdated.GetWorkflowExecutionOptionsUpdatedEventAttributes()
+				s.Nil(attributes.GetVersioningOverride())
+				s.False(attributes.GetUnsetVersioningOverride())
+
+				if tc.OnConflictOptions.AttachRequestId {
+					s.Equal(request.RequestId, attributes.GetAttachedRequestId())
+				} else {
+					s.Empty(attributes.GetAttachedRequestId())
+				}
+
+				if tc.OnConflictOptions.AttachCompletionCallbacks {
+					s.ProtoElementsMatch(request.CompletionCallbacks, attributes.GetAttachedCompletionCallbacks())
+				} else {
+					s.Empty(attributes.GetAttachedCompletionCallbacks())
+				}
+
+				if tc.OnConflictOptions.AttachLinks {
+					s.ProtoElementsMatch(request.Links, wfOptionsUpdated.GetLinks())
+				} else {
+					s.Empty(wfOptionsUpdated.GetLinks())
+				}
+			}
+
+			descResp, err := s.FrontendClient().DescribeWorkflowExecution(
+				testcore.NewContext(),
+				&workflowservice.DescribeWorkflowExecutionRequest{
+					Namespace: s.Namespace().String(),
+					Execution: &commonpb.WorkflowExecution{
+						WorkflowId: id,
+						RunId:      we0.RunId,
+					},
+				},
+			)
+			s.NoError(err)
+			s.Equal(enumspb.WORKFLOW_EXECUTION_STATUS_RUNNING, descResp.WorkflowExecutionInfo.Status)
+		})
+	}
+}
+
+func (s *WorkflowTestSuite) TestStartWorkflowExecution_UseExisting_OnConflictOptions_Dedup() {
+	id := "functional-start-workflow-use-existing-on-conflict-options-dedup-test"
+	wt := &commonpb.WorkflowType{
+		Name: "functional-start-workflow-use-existing-on-conflict-options-dedup-test-type",
+	}
+	tq := &taskqueuepb.TaskQueue{
+		Name: "functional-start-workflow-use-existing-on-conflict-options-dedup-test-taskqueue",
+		Kind: enumspb.TASK_QUEUE_KIND_NORMAL,
+	}
+
+	request := &workflowservice.StartWorkflowExecutionRequest{
+		RequestId:          uuid.New(),
+		Namespace:          s.Namespace().String(),
+		WorkflowId:         id,
+		WorkflowType:       wt,
+		TaskQueue:          tq,
+		Input:              nil,
+		WorkflowRunTimeout: durationpb.New(100 * time.Second),
+		Identity:           "worker1",
+	}
+
+	we0, err0 := s.FrontendClient().StartWorkflowExecution(testcore.NewContext(), request)
+	s.NoError(err0)
+	s.True(we0.Started)
+
+	request.RequestId = uuid.New()
+	request.WorkflowIdConflictPolicy = enumspb.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING
+	request.OnConflictOptions = &workflowpb.OnConflictOptions{
+		AttachRequestId: true,
+	}
+	we1, err1 := s.FrontendClient().StartWorkflowExecution(testcore.NewContext(), request)
+	s.NoError(err1)
+	s.Equal(we0.RunId, we1.RunId)
+	s.False(we1.Started)
+
+	we2, err2 := s.FrontendClient().StartWorkflowExecution(testcore.NewContext(), request)
```

Comment:
Worth doing another request without the OnConflictOptions to see that it gets deduped to for completeness.